File: insertion-sort/insertion-sort.js
const getIndex = i => Math.floor(Math.random() * i)

const generateArray = (size) => {
    let arr = []
    for (let i = 1; i <= size; i++) {
        const index = getIndex(i)
        arr.splice(index, 0, i)
    }
    return arr
}

const insertionSort = (arr) => {
    let res = []
    res[0] = arr[0]
    for (let i = 1; i < arr.length; i++) {
        let j = 0
        while (j < i && res[j] < arr[i]) j++
        if (j == res.length) {
            res.push(arr[i])
        } else {
            res.splice(j, 0, arr[i])
        }
    }
    return res
}

const main = () => {
    const size = 30_000

    const genStart = new Date().getTime()
    const arr = generateArray(size)
    const genEnd = new Date().getTime()

    console.log(`1. Time to generate ${genEnd - genStart} ms`)

    const sortStart = new Date().getTime()
    const res = insertionSort(arr)
    const sortEnd = new Date().getTime()

    const expected = []
    for (let i = 1; i <= arr.length; i++) expected.push(i)

    if (expected.toString() != res.toString()) {
        console.error("2. They are not equal")
    } else {
        console.log("2. Result array is as expected")
        console.log(`3. Sorted in ${sortEnd - sortStart} milliseconds`)
    }
}

main()

File: insertion-sort/insertion-sort.py
#! /usr/bin/env python

from math import floor
from random import random
import numpy as np
from time import time
from array import array

def get_index(i):
    return floor(random() * i)


def generate_array(size):
    arr = array("I", [])
    for i in range(1, size + 1):
        index = get_index(i)
        arr.insert(index, i)
    return arr


def insertion_sort(arr):
    res = array("I", [])
    res.append(arr[0])
    for i in range(1, len(arr)):
        elem = arr[i]      # This variable exists for performance
        j = 0
        while j < i and res[j] < elem:
            j += 1
        res.insert(j, elem)
    return res


def main():
    size = 30_000
    arr = generate_array(size) # Source Array (Scrambled)

    start = time()
    res = insertion_sort(arr) # Result Array
    end = time()

    expected = array("I", [])
    for i in range(1, size):
        expected.append(i)

    if (np.array_equal(res, expected)):
        print("Result arrays is NOT as expected")
    else:
        print("Result arrays is as expected")
        total = floor((end - start) * 1000)
        print(f"Sorted in {total} milliseconds")

main()
